lib/xwax/pitch.h

```diff
@@ -1,71 +1,157 @@
+ * Kalman filter replacement for alpha-beta pitch estimator
+ * Model: constant velocity (x,v) with acceleration process noise
  *
- * "xwax" is distributed in the hope that it will be useful, but
- * WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
- * General Public License for more details.
- *
- * You should have received a copy of the GNU General Public License
- * along with this program; if not, see <https://www.gnu.org/licenses/>.
+ * State:
+ *   x ... displacement over the last interval (kept relative by subtracting dx at the end)
+ *   v ... velocity (pitch)
  *
+ * Observation:
+ *   z = dx  (measured displacement over dt)
  */
 
 #ifndef PITCH_H
 #define PITCH_H
 
-/* Values for the filter concluded experimentally */
+#include <math.h>
+#include <string.h> /* for memset */
 
-#define ALPHA (1.0/512)
-#define BETA (ALPHA/256)
+struct pitch {
+    /* configuration */
+    double dt;     /* sample period (s) */
+    double q;      /* process noise spectral density (acceleration), e.g. (rad/s^2)^2 * s */
+    double r;      /* measurement noise variance for dx (units^2) */
 
-/* State of the pitch calculation filter */
+    /* state */
+    double x;      /* position (relative; we subtract dx each step) */
+    double v;      /* velocity (pitch) */
 
-struct pitch {
-    double dt, x, v;
+    /* covariance P (2x2, symmetric) */
+    double Pxx, Pxv, Pvv;
 };
 
-/* Prepare the filter for observations every dt seconds */
-
-static inline void pitch_init(struct pitch *p, double dt)
+/* Initialize with dt and noise params.
+   q: acceleration noise spectral density (tune up if motion is more jittery)
+   r: variance of dx measurement (tune up if observations are noisier) */
+static inline void pitch_init(struct pitch *p, double dt, double q, double r)
 {
     p->dt = dt;
+    p->q  = q;
+    p->r  = r;
+
     p->x = 0.0;
     p->v = 0.0;
-}
 
-/* Input an observation to the filter; in the last dt seconds the
- * position has moved by dx.
- *
- * Because the vinyl uses timestamps, the values for dx are discrete
- * rather than smooth. */
+    /* Start with large uncertainty so the filter trusts early measurements */
+    p->Pxx = 1e6;
+    p->Pxv = 0.0;
+    p->Pvv = 1e6;
+}
 
-static inline void pitch_dt_observation(struct pitch *p, double dx)
+/* Optionally retune noise without resetting state */
+static inline void pitch_set_noise(struct pitch *p, double q, double r)
 {
-    double predicted_x, predicted_v, residual_x;
-
-    predicted_x = p->x + p->v * p->dt;
-    predicted_v = p->v;
-
-    residual_x = dx - predicted_x;
-
-    p->x = predicted_x + residual_x * ALPHA;
-    p->v = predicted_v + residual_x * BETA / p->dt;
+    p->q = q;
+    p->r = r;
+}
 
-    p->x -= dx; /* relative to previous */
+/* Optionally change dt (if your cadence changes) */
+static inline void pitch_set_dt(struct pitch *p, double dt)
+{
+    p->dt = dt;
 }
```

Comment:
this seems to be unused. I guess this is supposed to be called on samplerate changes so the filter adapts without having to reinitialize the entire timecoder?

Reply:
Yes, usually `dt is also the wrong name in discrete time. This is just inherited from xwax. Usually `Ts = 1 / Fs` is the time step in this discrete domain `dt` for the continuous domain.

Reply:
Removed